src/librustc/ty/layout.rs

```diff
@@ -606,34 +609,219 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
+                    for local in fields {
+                        match assignments[*local] {
+                            Unassigned => {
+                                assignments[*local] = Assigned(variant_index);
+                            }
+                            Assigned(idx) => {
+                                // We've already seen this local at another suspension
+                                // point, so it is no longer a candidate.
+                                trace!("removing local {:?} in >1 variant ({:?}, {:?})",
+                                       local, variant_index, idx);
+                                eligible_locals.remove(*local);
+                                assignments[*local] = Ineligible(None);
+                            }
+                            Ineligible(_) => {},
+                        }
+                    }
+                }
+
+                // Next, check every pair of eligible locals to see if they
+                // conflict.
+                for (local_a, conflicts_a) in info.storage_conflicts.iter_enumerated() {
+                    if !eligible_locals.contains(local_a) {
+                        continue;
+                    }
+
+                    for local_b in conflicts_a.iter() {
+                        // local_a and local_b have overlapping storage, therefore they
+                        // cannot overlap in the generator layout. The only way to guarantee
+                        // this is if they are in the same variant, or one is ineligible
+                        // (which means it is stored in every variant).
+                        if !eligible_locals.contains(local_b) ||
+                            assignments[local_a] == assignments[local_b]
+                        {
+                            continue;
+                        }
+
+                        // If they conflict, we will choose one to make ineligible.
+                        let conflicts_b = &info.storage_conflicts[local_b];
+                        let (remove, other) = if conflicts_a.count() > conflicts_b.count() {
+                            (local_a, local_b)
+                        } else {
+                            (local_b, local_a)
+                        };
+                        eligible_locals.remove(remove);
+                        assignments[remove] = Ineligible(None);
+                        trace!("removing local {:?} due to conflict with {:?}", remove, other);
+                    }
+                }
+
+                let mut ineligible_locals = BitSet::new_filled(info.field_tys.len());
+                ineligible_locals.subtract(&eligible_locals);
 
+                // Write down the order of our locals that will be promoted to
+                // the prefix.
+                for (idx, local) in ineligible_locals.iter().enumerate() {
+                    assignments[local] = Ineligible(Some(idx as u32));
```

Comment:
How come this doesn't iterate over `assignments` instead?

Reply:
Then I would have to keep and increment my own counter for every local that happens to be assigned `Ineligible`.

Reply:
That would likely be more readable than bitset trickery, IMO.